From e1eba22e0aec95f755d4b93179683ad354bbb2fd Mon Sep 17 00:00:00 2001
From: wangrong <wangrong@uniontech.com>
Date: Tue, 4 Nov 2025 15:15:32 +0800
Subject: [PATCH] fix: fix read stuck when reenter mountpoint

When avfsd reads files, it acquires a global lock. If a request causes
avfsd to re-enter the mount point and attempt to read, it will cause a
second locking event, resulting in a block. The solution is to prevent
the fuse daemon from reading itself.

Log: fix read stuck when reenter mountpoint
Bug: https://pms.uniontech.com/bug-view-333583.html
---
Index: avfs-1.1.4/fuse/avfsd.c
===================================================================
--- avfs-1.1.4.orig/fuse/avfsd.c
+++ avfs-1.1.4/fuse/avfsd.c
@@ -24,6 +24,22 @@ struct fuse *fuse;
 
 static pthread_mutex_t avfsd_mutexlock = PTHREAD_MUTEX_INITIALIZER;
 
+static pid_t daemon_pid;
+
+static int avfsd_is_req_from_self()
+{
+    struct fuse_context *context;
+    char path[256];
+    struct stat st;
+
+    context = fuse_get_context();
+    if (daemon_pid == 0)
+        daemon_pid = getpid();
+    snprintf(path, sizeof(path), "/proc/%d/task/%d", daemon_pid, context->pid);
+
+    return (virt_lstat(path, &st) == 0) ? 1 : 0;
+}
+
 static int avfsd_getattr(const char *path, struct stat *stbuf)
 {
     int res;
@@ -214,6 +230,9 @@ static int avfsd_read(const char *path,
 
     (void) path;
 
+    if (avfsd_is_req_from_self())
+        return -EINVAL;
+
     pthread_mutex_lock( &avfsd_mutexlock );
     if (virt_lseek(fi->fh, offset, SEEK_SET) == -1) {
         pthread_mutex_unlock( &avfsd_mutexlock );
@@ -236,6 +255,9 @@ static int avfsd_write(const char *path,
 
     (void) path;
 
+    if (avfsd_is_req_from_self())
+        return -EINVAL;
+
     pthread_mutex_lock( &avfsd_mutexlock );
     if (virt_lseek(fi->fh, offset, SEEK_SET) == -1) {
         pthread_mutex_unlock( &avfsd_mutexlock );
@@ -255,6 +277,9 @@ static int avfsd_release(const char *pat
 {
     (void) path;
 
+    if (avfsd_is_req_from_self())
+        return -EINVAL;
+
     pthread_mutex_lock( &avfsd_mutexlock );
     virt_close(fi->fh);
     pthread_mutex_unlock( &avfsd_mutexlock );
